### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by wombach**
Merge branch performance improvement into master branch; this merge request is a requirement for merge request in voow-colos for branch add weather data

### CODE REVIEW

File: openclsim/core.py [hunk 1]
```diff
@@ -756,83 +772,96 @@ def __init__(self, criteria, *args, **kwargs):
         super().__init__(*args, **kwargs)
         """Initialization"""
         self.criteria = criteria
-        self.work_restrictions = {}
 
-    def calc_work_restrictions(self, location):
-        self.work_restrictions[location.name] = {}
+    def check_weather_restriction(self, location, event):
```
Comment (hoonhout) [inline openclsim/core.py:776]: Can we make this actually an abstract class?

File: openclsim/core.py [hunk 2]
```diff
@@ -932,7 +961,7 @@ def calc_depth_restrictions(self, location, processor):
                 seconds=processor.unloading(
                     self,
                     location,
-                    self.container.level - filling_degree * self.container.capacity,
+                    self.container.level + filling_degree * self.container.capacity,
```
Comment (hoonhout) [inline openclsim/core.py:964]: Really?

**Comment by wombach** [comment]
> Hi @wombach,
> 
> I get a few things from this PR:
> 
> * You improved the naming of variables (e.g. `mover` instead of `origin`).
> * You improved the modularity of methods (e.g. `delay_processing`).
> * You created abstract classes for the weather restrictions to test various methods.
> * You added `MoveActivity` for use in the updated `VOOW-COLOS` package.
> * You merged the more efficient method to compute workable time ranges by Mees and me.
> * You merged the method to define subcycles by Joris.
> 
> Let me know if I'm missing something. If not, these changes seem harmless to me.
> 
> I'm only wondering if we could rename the abstract classes rather than the actual classes for backwards compatibility; and make them actually abstract. But i'm not sure about a good name. `HasWorkabilityCriteria` and `HasWeather` are nicely descriptive for those base classes.

Hi @hoonhout ,

I used new names for the abstract classes to enable backward compatibility. All current implementations use the implementation class HasWorkabilityCriteria and HasWeather in their code. If I change these names all will break - although I agree that these would be better name for the abstract base classes. 
I will make the abstract class abstract base class.

File: openclsim/core.py (hunk 1 above)
Comment (wombach) [reply openclsim/core.py:785]: Change HasWeather to HasAbstractWeather and HasWorkabilityCriteria to HAsAbstractWorkabilityCriteria to be able to keep the original class names.
Further both classes do inherit now form ABC

File: openclsim/core.py (hunk 2 above)
Comment (wombach) [reply openclsim/core.py:973]: I can not comment on this since I have not touched this part of the code. Who did the change?

Comment (wombach) [reply openclsim/core.py]: If I turn it to minus the test fails. If it is a plus, the test passed. Thus, I kept it as a plus!

Comment (majstrooker) [reply openclsim/core.py:973]: Haha, I did not change it. If we really want to know you can check with the GitLens extension in VS Code.

**Pull request merged**